### PR TITLE
[9.0][ADD]sale_stock_picking_return_invoicing: Refund sale return pickings.

### PR DESCRIPTION
**Pull request by yajo**
Without this addon, when some customer returns some items to you and that
return picking is associated with a sale order, and then you invoice the sale
order, the returned items are never deducted from the total amount of delivered
products.

Also, if you already invoiced the sales order and again you get a refundable
return picking associated to it, you have no option to make a refund customer
invoice.

This addon enables both possibilities, by partially backporting this
functionality implementation from v10.

@Tecnativa

### CODE REVIEW

File: sale_stock_refund_picking/README.rst [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+Usage
+=====
+
+To use this module, when some customer returns some refundable products to you
+before you created an invoice, you need to:
+
+#. Go to *Sales > Sales Orders > Create*.
+#. Choose a customer and add a product whose *Invoicing Policy* is *Delivered
+   quantities*, and input that you are selling 20 units.
```
Comment (pedrobaeza) [inline sale_stock_refund_picking/README.rst]: Don't specify specific quantities, but talk in abstract: `and put some quantity to sell`, `return a lower quantity that the sold one`...

File: sale_stock_refund_picking/README.rst [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#. Confirm the sale.
+#. Go to *Delivery > Validate > Apply > Reverse*.
+#. Set *Quantity* to 5, and enable *To Refund*.
+#. Press *Return > Validate > Apply*.
+#. Return to the sale order.
+#. Press *Create Invoice > Invoiceable lines > Create and View Invoices*.
+#. The created invoice's amount is 15 units now.
```
Comment (pedrobaeza) [inline sale_stock_refund_picking/README.rst]: `The amount of the created invoice will be the difference between the sold and the returned quantity.`

File: sale_stock_refund_picking/README.rst [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/95/9.0
+
+Known issues / Roadmap
+======================
+
+* This addon is a pseudobackport of a functionality that exists natively in
+  v10, plus a fix for https://github.com/odoo/odoo/issues/13974, so this addon
```
Comment (pedrobaeza) [inline sale_stock_refund_picking/README.rst]: The issue is for purchase, so it doesn't apply here.

File: sale_stock_refund_picking/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Jairo Llopis <jairo.llopis@tecnativa.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+{
+    "name": "Refund Return Pickings in Sales Orders",
+    "summary": "Add an option to refund return pickings",
+    "version": "9.0.1.0.0",
+    "category": "Sales",
```
Comment (pedrobaeza) [inline sale_stock_refund_picking/__openerp__.py]: I think category is `Sales Management`

File: sale_stock_refund_picking/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    "website": "https://tecnativa.com/",
+    "author": "Tecnativa",
```
Comment (pedrobaeza) [inline sale_stock_refund_picking/__openerp__.py]: Add OCA

File: sale_stock_refund_picking/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    "license": "LGPL-3",
```
Comment (pedrobaeza) [inline sale_stock_refund_picking/__openerp__.py]: AGPL-3 for backend modules

File: sale_stock_refund_picking/models/sale_order.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    @api.multi
+    def _get_delivered_qty(self):
+        """Deduct moves marked to refund."""
+        qty = super(SaleOrderLine, self)._get_delivered_qty()
+        for move in self.procurement_ids.mapped('move_ids').filtered(
+                lambda r: (r.state == 'done' and
```
Comment (pedrobaeza) [inline sale_stock_refund_picking/models/sale_order.py]: Uhm, curious way to get refunds... This doesn't serve for purchases as returned doesn't preserve procurement group I think.

File: sale_stock_refund_picking/models/stock_move.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from openerp import fields, models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    to_refund_so = fields.Boolean(
+        "To Refund in SO",
```
Comment (pedrobaeza) [inline sale_stock_refund_picking/models/stock_move.py]: Better to call it `to_refund_sale`, as there's no need to put cryptic abbreviations.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Some cosmetic changes, but good in general

**Comment by pedrobaeza** [comment]
Please add a test for the second use case where the quantity delivered is negative.

**Comment by pedrobaeza** [comment]
About the name, I'm thinking that the word refund doesn't apply when subtracting quantity, so better to call it _sale_stock_picking_return_invoicing_.

File: sale_stock_refund_picking/README.rst (hunk 3 above)
Comment (yajo) [reply sale_stock_refund_picking/README.rst]: No, it applies, because it hides the field when the picking is not related to a purchase order; a different addon will implement this for POs.

I will tag the code as a HACK where it belongs, anyway.

File: sale_stock_refund_picking/__openerp__.py (hunk 4 above)
Comment (yajo) [reply sale_stock_refund_picking/__openerp__.py]: [Nope.](https://github.com/odoo/odoo/blob/9.0/addons/sale/__openerp__.py#L7)

File: sale_stock_refund_picking/models/sale_order.py (hunk 7 above)
Comment (yajo) [reply sale_stock_refund_picking/models/sale_order.py]: Honestly, I just copied that from v10.

File: sale_stock_refund_picking/models/stock_move.py (hunk 8 above)
Comment (yajo) [reply sale_stock_refund_picking/models/stock_move.py]: It's the field name in v10, kept it for that reason.

File: sale_stock_refund_picking/__openerp__.py (hunk 4 above)
Comment (pedrobaeza) [reply sale_stock_refund_picking/__openerp__.py]: OK

File: sale_stock_refund_picking/models/stock_move.py (hunk 8 above)
Comment (pedrobaeza) [reply sale_stock_refund_picking/models/stock_move.py]: OK, then let's go for the same, although as this is a TransientModel is more or less the same.

Comment (yajo) [reply sale_stock_refund_picking/models/stock_move.py]: Nice, just... it is a normal model. :stuck_out_tongue_winking_eye:

Comment (pedrobaeza) [reply sale_stock_refund_picking/models/stock_move.py]: Yeah, you're right... I mix the return wizard with this flag, but as said, don't change it.

**Comment by yajo** [comment]
All should be ready, except for a test that is failing, I think that's because of a bug in Odoo, [that I think has a wrong implementation for products to be sold by delivery](https://github.com/odoo/odoo/blob/9.0/addons/sale/sale.py#L482-L488). Me or @pedrobaeza will check that ASAP.

**Comment by yajo** [comment]
OK, after all Odoo was behaving alright.

I thought I had to check order state was `invoiced` after a return was issued, but it turns out that given the ordered qty is still bigger than the delivered and invoiced one (those 2 should be equal  BTW), status becomes `no`, and that's the expected behavior.

Just adapted the test and it should go :white_check_mark: now.

File: sale_stock_picking_return_invoicing/README.rst [hunk 9]
```diff
@@ -0,0 +1,97 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+======================================
+Refund Return Pickings in Sales Orders
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/README.rst]: Change to "Refund Returned Pickings from Sales Orders"

File: sale_stock_picking_return_invoicing/README.rst [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+======================================
+
+This module extends the functionality of sale orders to support marking some
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/README.rst]: s/sale order/sales orders invoicing

File: sale_stock_picking_return_invoicing/README.rst [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+stock picking moves as "To refund in Sale Order" and to allow you to create
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/README.rst]: s/stock picking moves/returned pickings

File: sale_stock_picking_return_invoicing/README.rst [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+To use this module, when some customer returns some refundable products to you
+before you created an invoice, you need to:
+
+#. Go to *Sales > Sales Orders > Create*.
+#. Choose a customer and add a product whose *Invoicing Policy* is *Delivered
+   quantities*, and input some quantity to sell.
+#. Confirm the sale.
+#. Go to *Delivery > Validate > Apply > Reverse*.
+#. Set *Quantity* to a lower quantity that the sold one, and enable
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/README.rst]: s/that/than

File: sale_stock_picking_return_invoicing/README.rst [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+   *To Refund*.
+#. Press *Return > Validate > Apply*.
+#. Return to the sale order.
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/README.rst]: s/Return/Go back

File: sale_stock_picking_return_invoicing/README.rst [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+#. Press *Create Invoice > Invoiceable lines > Create and View Invoices*.
+#. The amount of the created invoice will be the difference between the sold
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/README.rst]: s/sold/delivered

File: sale_stock_picking_return_invoicing/README.rst [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+   and the returned quantity.
+
+To use this module, when some customer returns some refundable products to you
+after you created an invoice, you need to:
+
+#. Go to *Sales > Sales Orders > Create*.
+#. Choose a customer and add a product whose *Invoicing Policy* is *Delivered
+   quantities*, and input that you are selling 20 units.
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/README.rst]: Put this also without specifying quantities

File: sale_stock_picking_return_invoicing/README.rst [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+#. Press *Create Invoice > Invoiceable lines > Create and View Invoices*.
+#. The created invoice's amount is 20 units now.
+#. Return to the sale order.
+#. Go to *Delivery > Reverse*.
+#. Set *Quantity* to 5, and enable *To Refund*.
+#. Press *Return > Validate > Apply*.
+#. Return to the sale order.
+#. Press *Create Invoice > Invoiceable lines (deduct down payments) >
+   Create and View Invoices*.
+#. The created invoice's amount is -5 units now.
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/README.rst]: This should be a refund invoice, not an invoice with -5.

File: sale_stock_picking_return_invoicing/models/stock_move.py [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+
+from openerp import fields, models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    to_refund_so = fields.Boolean(
+        "To Refund in SO",
+        default=False,
```
Comment (pedrobaeza) [inline sale_stock_picking_return_invoicing/models/stock_move.py]: This default is not need

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Documentation issues, but the rest LGTM

**Comment by pedrobaeza** [comment]
You have mixed all the modules/commits/branches in this PR.

**Comment by yajo** [comment]
Sorry, I don't know what did I do with git... Now it's OK, and I squashed everything.

**Review by rafaelbn — APPROVED** [review]
tested in runbot, i love your readme.

Great module

**Comment by rafaelbn** [comment]
@pedrobaeza take a look

**Review by pedrobaeza — APPROVED** [review]
